### PR TITLE
Refactor Linux strace

### PR DESCRIPTION
**Pull request by winson0123**
Previous implementation outputted strace logs onto the Linux guest VM. This refactor allows the strace logs to streamed directly to the CAPE result server on the host while the sample is running.

### CODE REVIEW

**Comment by kenleejl** [comment]
Great Job!!!! Thanks @winson0123 !

**Comment by BaconPotatoCat** [comment]
Wow!!! Amazing work! Thank you @winson0123

File: analyzer/linux/lib/common/results.py [hunk 1]
```diff
@@ -44,13 +44,41 @@ def upload_to_host(file_path, dump_path, pids="", ppids="", metadata="", categor
...
+    if config.upload_max_size < size + nc.buffer_size and not config.do_upload_max_size:
+        log.warning("Buffer is too big: %d; max size: %d", size + nc.buffer_size, config.upload_max_size)
+        return
+
+    try:
+        idx = 0
+        while idx < size:
+            nc.send(buffer[idx : idx + BUFSIZE], retry=True)
+            idx += BUFSIZE
+    except Exception as e:
```
Comment (nbargnesi) [inline analyzer/linux/lib/common/results.py:61]: Might be some room for improvement here - couple suggestions:

- Move the try/except around `nc.send()` where you're expecting exceptions
- If `nc.send()` does fail, I doubt `retry=True` is helpful and only a piece of the buffer isn't too useful. Maybe make this an error and return from the function?

File: analyzer/linux/lib/core/packages.py [hunk 2]
```diff
@@ -118,8 +119,11 @@ def start(self):
             # Remove the trailing slash (if any)
             self.target = filepath.rstrip("/")
         self.prepare()
-        # self.normal_analysis()
         self.strace_analysis()
+        self.nc.init("logs/strace.log", False)
+        self.thread = Thread(target=self.thread_send_strace_buffer)
+        self.thread.start()
```
Comment (nbargnesi) [inline analyzer/linux/lib/core/packages.py:125]: This thread will start immediately and potetenially miss `self.proc.stderr` when it runs. I think you want to synchroinize its execution with `strace_analysis`. Have a look at [threading event objects](https://docs.python.org/3/library/threading.html#event-objects). You can hold back the thread's execution until self.proc is readable.

And with self.proc.stderr being line-buffered I think - you can probably set a threading event each time a line is ready for reading.

**Comment by doomedraven** [comment]
this is fine to go right?

**Comment by winson0123** [comment]
Unless there's an issue with not passing the tests, I think it is good to go

**Comment by doomedraven** [comment]
thank you

**Pull request merged**